ui/src/waitingRoom.ts

```diff
@@ -0,0 +1,88 @@
+// Client representation of proxy waiting room requests, subscriptions, etc.
+import * as zod from "zod";
+
+export enum Status {
+  Created = "created",
+  Requested = "requested",
+  Found = "found",
+  Cloning = "cloning",
+  Cloned = "cloned",
+  Cancelled = "cancelled",
+  Failed = "failed",
+  TimedOut = "timedOut",
+}
+
+export interface ProjectRequest {
+  type: Status;
+  urn: string;
+}
+
+export enum PeerStatus {
+  Available = "available",
+  InProgress = "inProgress",
+  Failed = "failed",
+}
+
+export const eventSchema = zod.union([
+  zod.object({
+    type: zod.literal("tick"),
+  }),
+  zod.object({
+    type: zod.literal("created"),
+    urn: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("queried"),
+    urn: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("found"),
+    urn: zod.string(),
+    peer: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("cloning"),
+    urn: zod.string(),
+    peer: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("cloningFailed"),
+    urn: zod.string(),
+    peer: zod.string(),
+    reason: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("cloned"),
+    urn: zod.string(),
+    peer: zod.string(),
+  }),
+  zod.object({
+    type: zod.literal("canceled"),
+    urn: zod.string(),
+  }),
+]);
+
+export const roomStateSchema = zod.record(
+  zod.object({
+    state: zod.union([
+      zod.literal("Created"),
+      zod.literal("Requested"),
+      zod.literal("Found"),
+      zod.literal("Cloning"),
+      zod.literal("Cloned"),
+      zod.literal("Cancelled"),
+      zod.literal("Failed"),
+      zod.literal("TimedOut"),
+    ]),
+    peers: zod.record(
+      zod.union([
+        zod.literal("available"),
+        zod.literal("inProgress"),
+        zod.literal("failed"),
+      ])
+    ),
+  })
+);
+
+export type Event = zod.infer<typeof eventSchema>;
+export type RoomState = zod.infer<typeof roomStateSchema>;
```

Comment:
With this approach we don’t declare the `Event` and `RoomState` types explicitly. This makes it harder for developers to understand their shape (we have to look at the schema) and it makes it harder to debug typechecking errors. I’d suggest declaring these types explicitly
```typescript
type Event =
  {
    type: "created"
    urn: string,
 } | {
    // ...
  }
```

To support developers we can use explicit annotations for the schema
```typescript
const eventSchem: zod.Schema<Event> = ...
```

This is also the pattern we’re currently using in the rest of the codebase.